docs/TPPDialect.md

```diff
@@ -0,0 +1,355 @@
+  tpp.add ins(%arg0, %arg1) outs(%buf) : (memref<>, memref<>)
+  // In-of-place ops output on input operands
+  tpp.matmul ins(%A, %B, %C) outs(%C) : (memref<>, memref<>, memref<>)
+```
+
+### Unary
+
+On `tensors`, unary ops have one argument and one return value.
+
+On `memrefs`, unary operations have only `outs` for non-transforms (neg, inv, exp, etc.).
+For transforms (`broadcast`, `reduce`, `transpose`, `pack`, etc), there is one `ins` argument and one `outs` argument.
+The sizes must be the same for element-wise / arith, but will often be different for transforms (except `copy` and `transpose`).
+
+There is no implicit transform semantics.
+When lowering a broadcast (ex. a `linalg.generic` with lower dimension argument and apporpriate affine map), we map to `tpp.broadcast` + the operation.
+
+### Binary
+
+On `tensors`, binary ops have two arguments and one return value.
+
+On `memrefs`, binary operations have both `ins` and `outs`, with always two `ins` and one `outs` arguments.
+On most binary ops, the `outs` can alias `ins` (ex. `a=a+b`), and `ins` can alias themselves (ex. `a=a+a`).
+
+The exception is `matmul`, where the output cannot alias the inputs.
+A binary `matmul` is equivalent to `C = A x B`, with `C` as a new `memref` allocation.
+
+Element wise operations must have the same types, with broadcast done as a separate op, per operand.
+
+GEMM operations must have compatible types (MxNxK, VNNI, etc.).
+
+### Ternary
+
+On `tensors`, ternary ops have three arguments and one return value.
+
+On `memrefs`, ternary operations have both `ins` and `outs`, with always three `ins` and one `outs` arguments.
+The `outs` can alias with any type-compatible `ins` argument (accumulation matrix).
+
+A ternary `matmul` is equivalent to `C = A x B + C`, with `C` as an input and a potential output.
+
+GEMM operations must have compatible types (M,N,K) with the third input (C matrix) the same shape as the output.
+
+### Higher Order
+
+On higher order operations, both `ins` and `outs` arguments are mandatory, with more than three `ins` and multiple `outs` arguments.
+
+Tensor operations can return multiple results, but the bufferization reuse must follow the exact rules.
+
+These are _fused_ `brgemm` operations that can have multiple _fused_ arguments (bias add, relu) and return multiple values (intermediary results, for training).
+
+These ops are not implemented yet, but we need to keep the design space clear for them to exist.
+
+### Types
+
+TPP uses `tensor` and `memref` as its default data type.
+The supported element types are `fp32` and `bf16`.
+
+For advanced operations (ex. higher-order), we have the need to return special sparse tensors.
+We're still investigating what representation to use or if the existing sparse `tensor` attributes are enough.
+
+We may create additional types, as needed.
+For example:
+ * `vnni18`, `vnni8` for VNNI shapes of `bf16` and `bf8`. Those are _packed_ floats, ex. `{ bf16, bf16 }`.
```

Comment:
```suggestion
 * `vnni16`, `vnni8` for VNNI shapes of `bf16` and `bf8`. Those are _packed_ floats, ex. `{ bf16, bf16 }`.
```

Reply:
It also could be `vnni2` and `vnni4` based on the factors but I think it's better to keep 16 and 8 sizes equivalent to `bf` sizes.

Reply:
Hm, you're right. libxsmm uses `VNNI2` and `VNNI4` internally, so we should do the same.